pkg/registry/ecr/registry.go

```diff
@@ -0,0 +1,118 @@
+package ecr
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/v3io/registry-creds-handler/pkg/registry"
+	"github.com/v3io/registry-creds-handler/pkg/registry/abstract"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+)
+
+type Registry struct {
+	*abstract.Registry
+	awsCreds registry.AWSCreds
+}
+
+func NewRegistry(parentLogger logger.Logger,
+	secretName string,
+	namespace string,
+	creds string,
+	registryUri string) (*Registry, error) {
+	newRegistry := &Registry{}
+
+	// create base
+	abstractRegistry, err := abstract.NewRegistry(parentLogger, newRegistry, secretName, namespace, creds, registryUri)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create abstract registry")
+	}
+
+	newRegistry.Registry = abstractRegistry
+	return newRegistry, nil
+}
+
+func (r *Registry) EnrichAndValidate() error {
+	err := r.ValidateAbstractParams()
+	if err != nil {
+		return errors.Wrap(err, "Failed abstract registry params validation")
+	}
+
+	// parse aws credentials
+	var awsCreds registry.AWSCreds
+	if err := json.Unmarshal([]byte(r.Creds), &awsCreds); err != nil {
+		return errors.Wrap(err, "Failed to parse AWS credentials")
+	}
+
+	if awsCreds.Region == "" {
+		return errors.New("AWS Region is required")
+	}
+
+	if awsCreds.AccessKeyID == "" {
+		r.Logger.InfoWith("Did not receive AWS Access Key ID, checking env")
+		awsCreds.AccessKeyID = os.Getenv("AWS_ACCESS_KEY_ID")
```

Comment:
use a util such as

```golang
func GetFirstNonEmptyString(strings []string) string {
	for _, s := range strings {
		if s != "" {
			return s
		}
	}
	return ""
}
```

and then

create a ecr specific enrich function

```golang
awsCreds.SecretAccessKey = GetFirstNonEmptyString([]string{awsCreds.SecretAccessKey, os.Getenv("AWS_SECRET_ACCESS_KEY")})`

awsCreds. AccessKeyID = GetFirstNonEmptyString([]string{awsCreds. AccessKeyID, os.Getenv("AWS_ACCESS_KEY_ID")})
```

and ect specific validation function

```
...
if awsCreds.SecretAccessKey == "" {
  return errors.New("AWS Secret Access Key must not be empty")
}
...
```


and call those two functions. you will now have a neat EnrichAndValidate function